iwyu.cc

```diff
@@ -1650,13 +1676,34 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
       return;
 
     // Figure out the best location to attribute uses inside macros.
-    if (IsInMacro(used_loc))
-      used_loc = GetUseLocationForMacroExpansion(used_loc, decl);
     const FileEntry* used_in = GetFileEntry(used_loc);
+    const NamedDecl* fwd_decl = NULL;
+    bool isMacroDefinedAndUsedInSameFile = false;
+    if (IsInMacro(used_loc)) {
+      isMacroDefinedAndUsedInSameFile =
+          (GetFileEntry(used_loc) ==
+           GetFileEntry(GetInstantiationLoc(used_loc)));
+
+      const NamedDecl* fwd_decl = GetForwardDeclareInSameFile(decl, used_loc);
+      if (fwd_decl) {
+        // Make sure we keep the heuristic forward decl.
+        preprocessor_info().FileInfoFor(used_in)->ReportForwardDeclareUse(
+            used_loc, decl, IsNodeInsideCXXMethodBody(current_ast_node()),
+            comment);
+
+        // Attribute use to caller
+        used_loc = GetInstantiationLoc(used_loc);
+      } else {
+        used_loc = GetUseLocationForMacroExpansion(used_loc, decl);
+      }
```

Comment:
It is the same code as in `ReportDeclUseWithComment` but I don't know how to avoid code duplication. This code block requires `decl`, `used_loc`, returns `used_loc`, `fwd_decl` - looks like extracting it in a function will have too many arguments for input and output. Kim, have you tried this? Maybe it won't be so bad.